Monorepo: Enable lerna, add the root to the lerna config

blowery opened the pull request:
This allows us to symlink packages in `packages/*` from the root `node_modules`.

Testing Instructions:
* e2e tests should pass
* you should have a symlink from `node_modules/@automattic/tree-select` to the package in `packages/tree-select`

blowery committed:
Add the root to the lerna config

blowery committed:
use lerna bootstrap instead of npm ci

blowery committed:
add lockfiles, update shrinkwrap'

blowery commented on .circleci/config.yml:
```diff
@@ -115,7 +115,7 @@ references:
     command: |
       if [ ! -d node_modules ]; then
-        npm ci
+        npx lerna bootstrap
```
this is magic. lerna notices we're in circle and actually runs `npm ci` under the hood

sirreal replied:
Compared with `npm ci`, which will install packages according to lockfiles, lerna bootstrap should:
- run npm ci in all the packages by default in CI environment
- Symlink between any packages as need be

I don't think this caching scheme will be valid any longer 😞 

If the root packages.json and shrinkwrap are the same, we restore root node_modules and skip this command completely.

In a case where a packages was newly added and/or started to depend on another package in the repo, the cache would not be invalidated and the command would be skipped completely. That would not be the desired behavior.

We'll need to rethink how we cache this bit.

blowery commented on bin/install-if-deps-outdated.js:
```diff
@@ -24,7 +24,7 @@ const needsInstall = () => {
 
 if ( needsInstall() ) {
-	const installResult = spawnSync( 'npm', [ 'ci' ], {
+	const installResult = spawnSync( 'npx', [ 'lerna', 'bootstrap' ], {
```
should this also hand in `--ci`?

sirreal replied:
Before, we expected root shrinkwrap to exist, it seems that now we should expect package locks to exist and we could continue to use `CI`.
```suggestion
	const installResult = spawnSync( 'npx', [ 'lerna', 'bootstrap', '--ci' ], {
```

sirreal commented on Dockerfile:
```diff
@@ -26,6 +26,7 @@ RUN        bash /tmp/env-config.sh
 # Calypso development.
 COPY       . /calypso/
 RUN        npm ci --only=production --unsafe-perm
+RUN        npx lerna bootstrap --ci
```
Shall we just drop the preceding line?

sirreal commented on bin/install-if-deps-outdated.js:
```diff
@@ -24,7 +24,7 @@ const needsInstall = () => {
 };
 
 if ( needsInstall() ) {
```
It seems like `needsInstall` will need some reworking. It would need to take into account the root shrinkwrap as well as locks in any packages, right?

blowery replied:
reworked

sirreal commented on bin/install-if-no-packages.js:
```diff
@@ -3,7 +3,7 @@ const fs = require( 'fs' );
 if ( ! fs.existsSync( 'node_modules' ) ) {
 	console.log( 'No "node_modules" present, installing dependencies...' );
-	const installResult = spawnSync( 'npm', [ 'ci' ], {
+	const installResult = spawnSync( 'npx', [ 'lerna', 'bootstrap' ], {
```
Before, we expected root shrinkwrap to exist, it seems that now we should expect package locks to exist and we could continue to use `CI`.
```suggestion
	const installResult = spawnSync( 'npx', [ 'lerna', 'bootstrap', '--ci' ], {
```

sirreal commented on npm-shrinkwrap.json:
```diff
@@ -5581,7 +5581,7 @@
         },
```
Nothing relevant to this PR here (we can keep it, just noting that there are no package ref changes or similar)

sirreal committed:
Update bin/install-if-deps-outdated.js

Co-Authored-By: blowery <blowery@users.noreply.github.com>

sirreal committed:
Update bin/install-if-no-packages.js

Co-Authored-By: blowery <blowery@users.noreply.github.com>

blowery committed:
check node_modules in sub packages too

blowery committed:
simplify docker build

blowery committed:
update ignore changed

blowery committed:
Change back to the npm cache cache from #25487

sirreal commented on .circleci/config.yml:
```diff
@@ -93,35 +93,34 @@ references:
+    keys:
+      - v1-npm-modules-{{ checksum ".nvmrc" }}-{{ checksum "npm-shrinkwrap.json" }}
+      - v1-npm-modules-{{ checksum ".nvmrc" }}
+      - v1-npm-modules
```
I'm inclined to completely remove the cache when we change Node versions mostly to avoid having it grow indefinitely.

```suggestion
```

sirreal reviewed (commented):
One minor suggestion, but looks pretty good. I'll do some quick testing 👍

sirreal committed:
Remove fallback npm cache

sirreal reviewed (approved):
Seems to be working as expected 🙌

blowery committed:
make install script resilient to a missing node_modules
fix up distclean to leave locks in place
fix up update-deps to clean up lockfiles before re-running bootstrap

blowery committed:
interesting shrinkwrap.

blowery committed:
use spaces in tree-select package.json

blowery committed:
new shrinkwrap

simison commented on packages/tree-select/package.json:
```diff
@@ -1,31 +1,31 @@
+    ],
+    "author": "Ben Lowery <ben.lowery@automattic.com>",
+    "homepage": "https://github.com/Automattic/wp-calypso",
+    "license": "GPL-2.0-only",
```
Just curious, why not to align license with Calypso that is "GPLv2 or later"?

https://github.com/Automattic/wp-calypso/blob/0dd4cace13cd7065cfb782da90ebf686747ce375/package.json#L6

blowery replied:
no good reason, we should change it.

sirreal replied:
https://github.com/Automattic/wp-calypso/pull/29307
